docs/features/type-checking.md

```diff
@@ -0,0 +1,223 @@
+we narrow a type of `Person | Animal | None` using `hasattr(…, "name")`. `Person` is preserved in
+the narrowed union type because it has a `name` attribute, `Animal` is intersected with a synthetic
+protocol, accounting for the possibility of subclasses of `Animal` that add a `name` member. And
+`None` is excluded completely since it is a final type that has no `name` attribute:
+
+```py
+class Person:
+    name: str
+
+class Animal:
+    species: str
+
+def greet(being: Person | Animal | None):
+    if hasattr(being, "name"):
+        # `being` is now of type `Person | (Animal & <Protocol with members 'name'>)`
+
+        print(f"Hello, {being.name}!")
+    else:
+        print("Hello there!")
+```
+
+(Full example in the [playground](https://play.ty.dev/53173746-9676-4b63-b87a-88892f10a2a9))
+
+If you only use ty as a type checker, you can also make direct use of intersection types in
+annotations by importing `Intersection` from the special `ty_extensions` module that is (currently)
+only available at type-checking time:
+
+```py
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from ty_extensions import Intersection
+
+    type SerializableVersioned = Intersection[Serializable, Versioned]
+
+def output_as_json(obj: SerializableVersioned) -> str:
+    ...
+```
+
+(Full example in the [playground](https://play.ty.dev/f003e901-0e45-4f45-9759-d6db9d5e5f66))
+
+## Top and bottom materializations
+
+Gradual types generally have two special [materializations]. The top materialization represents the
+"largest" type that a gradual type can materialize to: the union of all possible materializations.
+For example, the top materialization of `Any` is `object`, and the top materialization of
+`Any & int` is `int`. For invariant generic classes, the top materialization cannot be expressed in
+Python's type system, but it is a useful type that ty intersects with when `isinstance` checks
+involve generic classes. For example, when checking `isinstance(…, list)`, ty intersects with the
+top materialization of `list[Unknown]`:
+
+```py
+class Item: ...
+
+def process(items: Item | list[Item]):
+    if isinstance(items, list):
+        # reveals: (Item & Top[list[Unknown]]) | list[Item]
+        reveal_type(items)
+```
+
+(Full example in the [playground](https://play.ty.dev/f6f28717-1020-4bc6-9365-2717af89b4c3))
+
+You might think that the type inside the `if` branch should just be `list[Item]`. What's happening
+instead is that ty tells us that `items` could also be a subclass of `Item` *and* `list`!
+
+While this is correct, you might still want to distinghuish between "actual" `Item`s and lists of
+`Item`s. You can achieve this either by making `Item` a `@final` class, or by doing the `isinstance`
+check against `Item` instead. In the latter case, the `else` branch will have a narrowed type of
+`list[Item] & ~Item`, which acts effectively like `list[Item]`.
```

Comment:
nit: consider changing the example to make `Item` `@final`, and adding an info box at the bottom about why it's there - it was a bit puzzling to read the code snippet, trying to work out what the type expression in the comment means before reading the answer below

Reply:
That's a great suggestion. Thank you.